### ISSUE TITLE
Supporting CHARMM-specific FLAGs in Amber prmtop file

### ISSUE THREAD

**Issue by charmm-gui** [openmm/openmm#3225]
We are working on supporting [SLipids FF](http://www.fos.su.se/~sasha/SLipids/) with Amber FFs.
SLipids FF is optimized to be used with Amber FFs, but it contains some parameters that are features of CHARMM FF.
For example, it adopts Urey-Bradley and harmonic improper energy terms which are features of CHARMM FF.
We have succeeded to generate Amber prmtop file with SLipids FF using CHARMM-specific FLAGs, but it appears that OpenMM does not support reading data in CHARMM-specific FLAGs.

https://github.com/openmm/openmm/blob/ad113a0cb37991a2de67a08026cf3b91616bafbe/wrappers/python/openmm/app/internal/amber_file_parser.py#L131-L134

It would be great if OpenMM can read data in CHARMM-specific FLAGs.

Followings are CHARMM-specific FLAGs:
CHARMM_UREY_BRADLEY_COUNT
CHARMM_UREY_BRADLEY
CHARMM_UREY_BRADLEY_FORCE_CONSTANT
CHARMM_UREY_BRADLEY_EQUIL_VALUE
CHARMM_NUM_IMPROPERS
CHARMM_IMPROPERS
CHARMM_NUM_IMPR_TYPES
CHARMM_IMPROPER_FORCE_CONSTANT
CHARMM_IMPROPER_PHASE
LENNARD_JONES_14_ACOEF
LENNARD_JONES_14_BCOEF

The example prmtop file is attached.
[SLipids_example.tar.gz](https://github.com/openmm/openmm/files/7062745/SLipids_example.tar.gz)

**Comment by jchodera** [comment]
Would it make sense to use ParmEd for this, since it supports CHAMBER-style prmtop files?

**Comment by charmm-gui** [comment]
We don't want to add any dependency to the input provided by the CHARMM-GUI unless ParmEd becomes a standard file I/O protocol in OpenMM.

I have attached example prmtop and rst files: [slipids.tar.gz](https://github.com/openmm/openmm/files/7204277/slipids.tar.gz)

The energies of each component calculated by AMBER are (unit: kcal/mol):
Bond: 27,298.2060
Angle: 33,883.8347
Urey-Bradly: 12,187.4028
Dihedral: 3,220.6655
Improper: 328.5880
vdW 1-4: 4,209.1512
vdW: 48,174.1758
Elec 1-4: 5,747.5765
Elec: -346,295.3628

with 9-Å hard truncation cut-off for both vdW and electrostatic without PME.

**Comment by charmm-gui** [comment]
We found a minor conversion issue for SLipids and fixed it.
Please use this example instead: [slipids.tar.gz](https://github.com/openmm/openmm/files/7505960/slipids.tar.gz)

Term | Energy (unit: kcal/mol)
-- | --
Bond | 28,554.9842
Angle | 34,037.0175
Urey-Bradly | 12,232.6323
Dihedral | 3,227.4146
Improper | 330.6573
vdW 1-4 | 4,236.3610
vdW | 48,208.6402
Elec. 1-4 | 5,747.4813
Elec. | -346,287.7369

**Comment by alexandrepintochem** [comment]
> We found a minor conversion issue for SLipids and fixed it. Please use this example instead: [slipids.tar.gz](https://github.com/openmm/openmm/files/7505960/slipids.tar.gz)
> 
> Term	Energy (unit: kcal/mol)
> Bond	28,554.9842
> Angle	34,037.0175
> Urey-Bradly	12,232.6323
> Dihedral	3,227.4146
> Improper	330.6573
> vdW 1-4	4,236.3610
> vdW	48,208.6402
> Elec. 1-4	5,747.4813
> Elec.	-346,287.7369

How did you fix the conversion for Slipids? I've been trying to convert GROMACS .top and .gro (Slipids bilayer + Protein) to AMBER .prmtop and .inpcrd with parmed, so that I can load the system to CP2K. However, the CHARMM style .prmtop generated seems to be incompatible with CP2K (for example, I have to remove all the %COMMENT lines so that the file is read, but then it doesn't detect the Urey-Bradley parameters).

**Comment by charmm-gui** [comment]
@alexandrepintochem If CP2K does not load CHABMER style prmtop file properly, then it is problem of CP2K. You'd better to contact CP2K developer.

**Comment by alexandrepintochem** [comment]
> @alexandrepintochem If CP2K does not load CHABMER style prmtop file properly, then it is problem of CP2K. You'd better to contact CP2K developer.

I was asking how did you get the Slipids AMBER parameter file with Urey-Bradley parameters without being CHARMM style? CP2K supports Urey-Bradley parameters and nomal AMBER parameter files, so that would solve it for me. Sorry if my question seems misplaced, but it would be more helpful to me knowing how can I make a regular AMBER .prmtop for Slipids, rather than ask the CP2K developers to support CHARMM style .prmtop files (altough I can also report this to the CP2K forum).

**Comment by peastman** [comment]
The file doesn't seem to match the documentation.  For the CHARMM_UREY_BRADLEY section it says, "It is formatted exactly like BONDS_INC_HYDROGEN and BONDS_WITHOUT_HYDROGEN."  For those two sections, it says the first two entries for each bond are indices into a coordinate array, that is, the atom index times 3.  That appears to be the case in the file:

```
%FLAG BONDS_INC_HYDROGEN
%FORMAT(10I8)
       3       0      20       6       0      20       9       0      20      15
```

All the indices are multiples of three.  But for the CHARMM_UREY_BRADLEY section they aren't:

```
%FLAG CHARMM_UREY_BRADLEY
%FORMAT(10I8)
       2       3      26       2       4      26       2       5      25       3
```

Can you confirm this is correct, and the first two elements are the indices of atoms, not coordinates?  If so, are they 0 or 1 based?

**Comment by charmm-gui** [comment]
Those are indices of atoms starting from 1.

**Comment by peastman** [comment]
Thanks!

I have the improper and Urey-Bradley terms matching the energies above.  I can't get any of the nonbonded terms to match though.  I'd like to try converting some smaller systems and matching up individual interactions.  How do I get chamber?  I installed AmberTools 21.11 with conda, but it doesn't seem to be there.  I do have other programs from AmberTools like tleap and antechamber, so I'm pretty sure I installed it correctly.  Just no chamber...

**Comment by charmm-gui** [comment]
chamber is removed from the default compilation list in AmberTools.
You may need to compile it separately using "make chamber", if I remember correctly.

**Comment by charmm-gui** [comment]
For your information, the non bonded energy above is calculated using the 9-Å hard truncation cut-off.
As far as I know, OpenMM does not support it. Correct?

**Comment by peastman** [comment]
Thanks!  I found that we already have a chamber converted file in wrappers/python/tests/systems.  Previously we were just checking to make sure AmberPrmtopFile threw an exception on it.  Now I'm trying to check that we actually get correct energies, comparing to ones computed with `pytraj.energy_decomposition()`.  I think I have the nonbonded terms agreeing.  Now I just need to figure out why the CMAP is missing.

**Comment by peastman** [comment]
You didn't mention anything about supporting CMAP, but I assume that's needed also?  The existing code looks for flags like `CMAP_COUNT` and `CMAP_RESOLUTION`.  Those aren't mentioned in https://ambermd.org/prmtop.pdf.  It lists them as `CHARMM_CMAP_COUNT`, `CHARMM_CMAP_RESOLUTION`, etc.  I assume the documentation is out of date, and when Amber19 added CMAPs, they started supporting those flags without the `CHARMM_` prefix.

Can I assume both are equivalent and I should just make it accept all flags with or without the prefix?

**Comment by peastman** [comment]
The implementation is in #3416.  Can you take a look?

**Comment by charmm-gui** [comment]
> Can I assume both are equivalent and I should just make it accept all flags with or without the prefix?

Yes, they are equivalent and Amber accepts both flags.

> The implementation is in #3416. Can you take a look?

Thanks!! I will check it and get back to you shortly.

**Comment by charmm-gui** [comment]
There is one issue in handling 1-4 interaction.
`SCEE_SCALE_FACTOR` and `SCNB_SCALE_FACTOR` scales 1-4 non-bonded energy calculated using the parameters in `LENNARD_JONES_14_ACOEF` and `LENNARD_JONES_14_BCOEF`.
Therefore, following codes in `get14Interactions()`:
```
if self.chamber:
    iScee = 1.0
    iScnb = 1.0
else:
    try:
        iScee = float(self._raw_data['SCEE_SCALE_FACTOR'][iidx])
    except KeyError:
        iScee = 1.2
    try:
        iScnb = float(self._raw_data['SCNB_SCALE_FACTOR'][iidx])
    except KeyError:
        iScnb = 2.0
```
needs to be changed to:
```
try:
    iScee = float(self._raw_data['SCEE_SCALE_FACTOR'][iidx])
except KeyError:
    iScee = 1.2
try:
    iScnb = float(self._raw_data['SCNB_SCALE_FACTOR'][iidx])
except KeyError:
    iScnb = 2.0
```

**Comment by peastman** [comment]
If `chamber` is set, shouldn't both scale factors default to 1?  That is, should it be

```python
try:
    iScee = float(self._raw_data['SCEE_SCALE_FACTOR'][iidx])
except KeyError:
    iScee = 1.0 if self.chamber else 1.2
try:
    iScnb = float(self._raw_data['SCNB_SCALE_FACTOR'][iidx])
except KeyError:
    iScnb = 1.0 if self.chamber else 2.0
```

**Comment by charmm-gui** [comment]
I got your point.
Yes, that makes sense although I don't think that there could be any case where 'SCEE_SCALE_FACTOR' or 'SCNB_SCALE_FACTOR' is not in prmtop file.

**Comment by peastman** [comment]
Implemented in #3416.
